File: javascript/discordjs-core/src/bot.ts
import { DatabaseSync } from "node:sqlite";
import {
  ActivityType,
  ApplicationCommandOptionType,
  ApplicationCommandType,
  ApplicationIntegrationType,
  ChannelType,
  Client,
  GatewayDispatchEvents,
  GatewayIntentBits,
  InteractionContextType,
  InteractionType,
  MessageFlags,
  PermissionFlagsBits,
  PresenceUpdateStatus,
  type APIChatInputApplicationCommandInteractionData,
} from "@discordjs/core";
import { REST } from "@discordjs/rest";
import { WebSocketManager } from "@discordjs/ws";

const token = process.env.DISCORD_TOKEN;
if (!token) throw new Error("DISCORD_TOKEN is required");

const db = new DatabaseSync(process.env.DB_PATH ?? "bot.db");
const sql = db.createTagStore();

process.on('uncaughtException', (err) => {
  console.error(`Unhandled Exception: ${err}`);
});

process.on('unhandledRejection', (reason, promise) => {
  console.error(`Unhandled Rejection at: ${promise}, reason: ${reason}`);
});

const rest = new REST({ version: "10" }).setToken(token);
const gateway = new WebSocketManager({
  token,
  // Guilds + guild messages are the only intents the honeypot needs.
  intents: GatewayIntentBits.Guilds | GatewayIntentBits.GuildMessages,
  rest,
  initialPresence: {
    status: PresenceUpdateStatus.Online,
    since: null,
    afk: false,
    activities: [{
      name: "#honeypot",
      state: "Watching #honeypot for bots",
      type: ActivityType.Custom,
    }],
  },
});
const client = new Client({ rest, gateway });

// Guild honeypot config is stored in SQLite so it survives restarts.
sql.run`
  CREATE TABLE IF NOT EXISTS honeypots (
    guild_id TEXT PRIMARY KEY,
    channel_id TEXT NOT NULL,
    action TEXT NOT NULL DEFAULT 'ban',
    log_channel_id TEXT
  )
`;

client.once(GatewayDispatchEvents.Ready, async ({ data, api }) => {
  console.log(`Logged in as ${data.user.username}#${data.user.discriminator}`);

  // Register slash commands once on startup.
  await api.applicationCommands.bulkOverwriteGlobalCommands(data.user.id, [
    {
      name: "honeypot-set",
      description: "Set/update honeypot channel (note: this overrides previous config set)",
      default_member_permissions: String(PermissionFlagsBits.BanMembers | PermissionFlagsBits.ManageChannels),
      contexts: [InteractionContextType.Guild],
      integration_types: [ApplicationIntegrationType.GuildInstall],
      options: [
        {
          type: ApplicationCommandOptionType.Channel,
          name: "channel",
          description: "The channel to ban people that message in it",
          required: true,
          channel_types: [ChannelType.GuildText],
        },
        {
          type: ApplicationCommandOptionType.String,
          name: "action",
          description: "The action to take when someone messages in the honeypot channel",
          required: true,
          choices: [
            { name: "Ban", value: "ban" },
            { name: "Softban", value: "softban" },
            { name: "Disabled", value: "disabled" },
          ],
        },
        {
          type: ApplicationCommandOptionType.Channel,
          name: "log_channel",
          description: "The channel to log actions in (if ommited, then it won't log anywhere)",
          required: false,
          channel_types: [ChannelType.GuildText, ChannelType.PublicThread, ChannelType.PrivateThread],
        },
      ],
    },
  ]);
});

function getOptionValue<R extends boolean | undefined>
  (data: APIChatInputApplicationCommandInteractionData, name: string, required: R = false as R)
  : R extends true ? string : string | undefined {
  const option = data.options?.find((o) => o.name === name);
  if (!option) {
    if (required) throw new Error(`Missing required option: ${name}`);
    return undefined as any;
  }
  return "value" in option ? String(option.value) : undefined as any;
}

function hasPermission(memberPermissions: string | undefined, required: bigint): boolean {
  return (BigInt(memberPermissions ?? 0) & required) === required;
}

// Handle the /honeypot-set command.
client.on(GatewayDispatchEvents.InteractionCreate, async ({ data: interaction, api }) => {
  if (interaction.type !== InteractionType.ApplicationCommand || !interaction.guild_id) return;
  if (interaction.data.type !== ApplicationCommandType.ChatInput) return;

  if (interaction.data.name === "honeypot-set") {
    const required = PermissionFlagsBits.BanMembers | PermissionFlagsBits.ManageChannels;
    if (!hasPermission(interaction.member?.permissions, required)) {
      await api.interactions.reply(interaction.id, interaction.token, {
        content: "You don't have permission to use this command.",
        flags: MessageFlags.Ephemeral,
      });
      return;
    }

    const channelId = getOptionValue(interaction.data, "channel", true);
    const action = getOptionValue(interaction.data, "action", true);
    const logChannelId = getOptionValue(interaction.data, "log_channel");

    if (action === "disabled") {
      // "disabled" removes the config instead of updating it.
      sql.run`DELETE FROM honeypots WHERE guild_id = ${interaction.guild_id}`;

      await api.interactions.reply(interaction.id, interaction.token, {
        content: "Honeypot configuration updated: Disabled honeypot for this server.",
      });
      return;
    }

    sql.run`
      INSERT INTO honeypots (guild_id, channel_id, action, log_channel_id)
      VALUES (${interaction.guild_id}, ${channelId}, ${action}, ${logChannelId ?? null})
      ON CONFLICT(guild_id) DO UPDATE SET
        channel_id = excluded.channel_id,
        action = excluded.action,
        log_channel_id = excluded.log_channel_id
    `;

    await api.interactions.reply(interaction.id, interaction.token, {
      content: `Honeypot configuration updated: Will **${action}** anyone who types in <#${channelId}> ${logChannelId ? `and log actions to <#${logChannelId}>` : "and won't log actions"}.`,
    });
  } else {
    await api.interactions.reply(interaction.id, interaction.token, {
      content: "Unknown command",
      flags: MessageFlags.Ephemeral,
    });
  }
});

interface Honeypot {
  channel_id: string;
  action: "ban" | "softban";
  log_channel_id: string | null;
}

// Ban anyone who messages in the configured honeypot channel.
client.on(GatewayDispatchEvents.MessageCreate, async ({ data: message, api }) => {
  if (!message.guild_id || message.author.bot) return;

  const honeypot = sql.get`SELECT * FROM honeypots WHERE guild_id = ${message.guild_id}` as Honeypot | undefined;

  if (!honeypot || honeypot.channel_id !== message.channel_id) return;

  let success = true;
  try {
    if (honeypot.action === "ban") {
      await api.guilds.banUser(message.guild_id, message.author.id, {
        delete_message_seconds: 3600, // 1hr
      }, {
        reason: "User typed in #honeypot channel -> ban",
      });
    } else if (honeypot.action === "softban") {
      await api.guilds.banUser(message.guild_id, message.author.id, {
        delete_message_seconds: 3600, // 1hr
      }, {
        reason: "User typed in #honeypot channel -> softban (1/2)",
      });
      await api.guilds.unbanUser(message.guild_id, message.author.id, {
        reason: "User typed in #honeypot channel -> softban (2/2)",
      });
    } else {
      throw new Error(`Unknown honeypot action: ${honeypot.action}`);
    }
  } catch (error) {
    success = false;
    console.error(`Failed honeypot action (${honeypot.action}): ${error}`);
  }

  // On failure, log to the honeypot channel itself so moderators still see it.
  const targetChannelId = success
    ? honeypot.log_channel_id
    : honeypot.log_channel_id ?? message.channel_id;
  if (!targetChannelId) return;

  // Send the log message directly — no need to fetch the target channel first.
  try {
    await api.channels.createMessage(targetChannelId, {
      content: success
        ? `User <@${message.author.id}> was ${honeypot.action} for triggering the honeypot in <#${message.channel_id}>`
        : `User <@${message.author.id}> triggered the honeypot but I **failed** to ${honeypot.action} them, please check my permissions to ensure I can ${honeypot.action} them.`,
      allowed_mentions: {},
    });
  } catch (error) {
    console.error(`Failed to send honeypot log message: ${error}`);
  }
});

// Clean up config when a channel or guild goes away.
client.on(GatewayDispatchEvents.ChannelDelete, ({ data }) => {
  sql.run`DELETE FROM honeypots WHERE channel_id = ${data.id}`;
});
client.on(GatewayDispatchEvents.GuildDelete, ({ data }) => {
  sql.run`DELETE FROM honeypots WHERE guild_id = ${data.id}`;
});

gateway.connect();

File: python/discord.py/bot.py
from __future__ import annotations

import os
import sqlite3
import sys
from typing import Optional

import discord
from discord import app_commands

TOKEN = os.environ.get("DISCORD_TOKEN")
if not TOKEN:
    raise RuntimeError("DISCORD_TOKEN is required")

DB_PATH = os.environ.get("DB_PATH", "bot.db")


class ChannelTransformer(app_commands.Transformer):
    def __init__(self, *types: discord.ChannelType) -> None:
        self._types = list(types)

    @property
    def type(self) -> discord.AppCommandOptionType:
        return discord.AppCommandOptionType.channel

    @property
    def channel_types(self) -> list[discord.ChannelType]:
        return self._types

    async def transform(self, interaction: discord.Interaction, value, /):
        return value


class HoneypotClient(discord.Client):
    def __init__(self) -> None:
        super().__init__(
            intents=discord.Intents(guilds=True, messages=True),
            activity=discord.CustomActivity(name="Watching #honeypot for bots"),
            status=discord.Status.online,
        )
        self.tree = app_commands.CommandTree(self)

        self.db = sqlite3.connect(DB_PATH)
        self.db.row_factory = sqlite3.Row
        self.db.execute(
            """
            CREATE TABLE IF NOT EXISTS honeypots (
                guild_id TEXT PRIMARY KEY,
                channel_id TEXT NOT NULL,
                action TEXT NOT NULL DEFAULT 'ban',
                log_channel_id TEXT
            )
            """
        )
        self.db.commit()

        self._register_commands()

    def _register_commands(self) -> None:
        @self.tree.command(
            name="honeypot-set",
            description="Set/update honeypot channel (note: this overrides previous config set)",
        )
        @app_commands.describe(
            channel="The channel to ban people that message in it",
            action="The action to take when someone messages in the honeypot channel",
            log_channel="The channel to log actions in (if ommited, then it won't log anywhere)",
        )
        @app_commands.choices(
            action=[
                app_commands.Choice(name="Ban", value="ban"),
                app_commands.Choice(name="Softban", value="softban"),
                app_commands.Choice(name="Disabled", value="disabled"),
            ]
        )
        @app_commands.guild_only()
        @app_commands.guild_install()
        @app_commands.default_permissions(ban_members=True, manage_channels=True)
        async def honeypot_set(
            interaction: discord.Interaction,
            channel: app_commands.Transform[
                discord.TextChannel,
                ChannelTransformer(discord.ChannelType.text),
            ],
            action: str,
            log_channel: Optional[
                app_commands.Transform[
                    discord.TextChannel,
                    ChannelTransformer(
                        discord.ChannelType.text,
                        discord.ChannelType.public_thread,
                        discord.ChannelType.private_thread,
                    ),
                ]
            ] = None,
        ) -> None:
            permissions = interaction.user.guild_permissions
            if not (permissions.ban_members and permissions.manage_channels):
                await interaction.response.send_message(
                    "You don't have permission to use this command.", ephemeral=True
                )
                return

            if action == "disabled":
                self.db.execute(
                    "DELETE FROM honeypots WHERE guild_id = ?",
                    (str(interaction.guild_id),),
                )
                self.db.commit()
                await interaction.response.send_message(
                    "Honeypot configuration updated: Disabled honeypot for this server."
                )
                return

            self.db.execute(
                """
                INSERT INTO honeypots (guild_id, channel_id, action, log_channel_id)
                VALUES (?, ?, ?, ?)
                ON CONFLICT(guild_id) DO UPDATE SET
                    channel_id = excluded.channel_id,
                    action = excluded.action,
                    log_channel_id = excluded.log_channel_id
                """,
                (
                    str(interaction.guild_id),
                    str(channel.id),
                    action,
                    str(log_channel.id) if log_channel else None,
                ),
            )
            self.db.commit()

            log_suffix = (
                f"and log actions to <#{log_channel.id}>"
                if log_channel
                else "and won't log actions"
            )
            await interaction.response.send_message(
                f"Honeypot configuration updated: Will **{action}** anyone who types in "
                f"<#{channel.id}> {log_suffix}."
            )

    async def on_ready(self) -> None:
        print(f"Logged in as {self.user}", flush=True)
        await self.tree.sync()

    async def on_message(self, message: discord.Message) -> None:
        if message.author.bot or not message.guild:
            return

        row = self.db.execute(
            "SELECT * FROM honeypots WHERE guild_id = ?", (str(message.guild.id),)
        ).fetchone()
        if row is None or row["channel_id"] != str(message.channel.id):
            return

        action = row["action"]
        success = True
        try:
            if action == "ban":
                await message.author.ban(
                    delete_message_seconds=3600,
                    reason="User typed in #honeypot channel -> ban",
                )
            elif action == "softban":
                await message.author.ban(
                    delete_message_seconds=3600,
                    reason="User typed in #honeypot channel -> softban (1/2)",
                )
                await message.guild.unban(
                    message.author.id,
                    reason="User typed in #honeypot channel -> softban (2/2)",
                )
            else:
                raise RuntimeError(f"Unknown honeypot action: {action}")
        except Exception as error:
            success = False
            print(f"Failed honeypot action ({action}): {error}", file=sys.stderr)

        target_channel_id = (
            row["log_channel_id"]
            if success
            else row["log_channel_id"] or str(message.channel.id)
        )
        if target_channel_id is None:
            return

        target_channel = await message.guild.fetch_channel(int(target_channel_id))
        if not isinstance(target_channel, discord.abc.Messageable):
            return

        try:
            await target_channel.send(
                content=(
                    f"User <@{message.author.id}> was {action} for triggering the honeypot "
                    f"in <#{message.channel.id}>"
                    if success
                    else f"User <@{message.author.id}> triggered the honeypot but I "
                    f"**failed** to {action} them, please check my permissions to "
                    f"ensure I can {action} them."
                ),
                allowed_mentions=discord.AllowedMentions.none(),
            )
        except Exception as error:
            print(f"Failed to send honeypot log message: {error}", file=sys.stderr)

    async def on_guild_channel_delete(self, channel: discord.abc.GuildChannel) -> None:
        self.db.execute(
            "DELETE FROM honeypots WHERE channel_id = ?", (str(channel.id),)
        )
        self.db.commit()

    async def on_guild_remove(self, guild: discord.Guild) -> None:
        self.db.execute(
            "DELETE FROM honeypots WHERE guild_id = ?", (str(guild.id),)
        )
        self.db.commit()


def main() -> None:
    client = HoneypotClient()
    client.run(TOKEN)


if __name__ == "__main__":
    main()
